src/urllib3/poolmanager.py

```diff
@@ -168,7 +168,7 @@ def _default_key_normalizer(
 pool_classes_by_scheme = {"http": HTTPConnectionPool, "https": HTTPSConnectionPool}
 
 
-class PoolManager(RequestMethods):
+class PoolManager(RequestMethods[HTTPResponse]):
```

Comment:
This one I'm not sure about -- if both users of `RequestMethods` use `HTTPResponse` then why does `BaseHTTPResponse` exist?

Looking into `BaseHTTPResponse` I found #1986 where @sethmlarson mentions

> Ensure that type hints for .request() of BaseHTTPConnection annotate as BaseHTTPResponse instead of HTTPResponse

but currently there is no `BaseHTTPConnection` so I'm guessing this work is still in progress?

Reply:
The reason we want to have `BaseHTTPConnection` and `BaseHTTPResponse` is so we can decouple ourselves from the httplib APIs that we're not guaranteeing to be available for alternate HTTP protocol implementations. So for PoolManager and ConnectionPool we shouldn't be typing the responses/connections to be `HTTPResponse`, they should be `BaseHTTPResponse` so users know which APIs will be supported into the future.

Does this make sense? If not we can revisit this, the primary goal is to decouple our API from httplib.

Reply:
Makes sense.

The relevant point for this PR is how does `PoolManager` fit into this:

- Will be used as is with some configuration which will decide which backend `request` etc. use.
    - Configuration is dynamic: cannot make `PoolManager` generic in the backend.
    - Configuration is static (defined when the `PoolManager` is constructed): can use a generic.
- Will have `BasePoolManager` with subclasses for each backend.
- Will have different methods for different backends.
- Or maybe the library should be designed such that the backend never needs to "leak" i.e. all functionality should be present on the base types and the sub-types never need to be exposed in any public API.
- Something else

If this is not decided yet, then we should be careful not to add typing that will lock a specific approach, i.e. probably defer these parts of the PR.

Reply:
@bluetech I think the fourth item you listed is what we're aiming for now:

> library should be designed such that the backend never needs to "leak" i.e. all functionality should be present on the base types and the sub-types never need to be exposed in any public API.

I'm not sure if generics are needed for this to work out fine, I'm really happy someone is putting extra brain cycles on this problem though.